### PR TITLE
#513 Log to track which tools provided data

### PR DESCRIPTION
**Pull request by kartiksibal**
Signed Off By: Kartik Sibal <kartiksibal@gmail.com>

### CODE REVIEW

File: src/scancode/cli.py [hunk 1]
```diff
@@ -442,7 +443,14 @@ def scan_event(item):
 
     files_scanned_per_second = round(float(files_count) / scanning_time , 2)
     scan_summary['files_scanned_per_second'] = files_scanned_per_second
-
+    
+    if not scanning_errors:
+        #Log file containing timestamp & Project's name.
+        log = open (input_path + "_log.txt", "a+")
```
Comment (pombredanne) [inline src/scancode/cli.py]: you should use a `with` statement (to close the file). You are leaking an open file

File: src/scancode/cli.py [hunk 2]
```diff
@@ -442,7 +443,14 @@ def scan_event(item):
...
+        day = datetime.datetime.today().replace(microsecond=0)
```
Comment (pombredanne) [inline src/scancode/cli.py]: Check commoncode. There is likley some utility function there or elswehere to get a timestamp as a string

File: src/scancode/cli.py [hunk 3]
```diff
@@ -34,6 +34,7 @@
 from collections import OrderedDict
 from functools import partial
 import os
+import datetime
```
Comment (pombredanne) [inline src/scancode/cli.py]: we like imports to be grouped, then sorted. This should go right after the collections import

**Comment by pombredanne** [comment]
Thanks! I provided some comments for you consideration.

File: src/scancode/cli.py (hunk 3 above)
Comment (kartiksibal) [inline src/scancode/cli.py]: Noted.

File: src/scancode/cli.py (hunk 1 above)
Comment (kartiksibal) [inline src/scancode/cli.py]: That was really stupid of me, sorry.

File: src/scancode/cli.py (hunk 2 above)
Comment (kartiksibal) [inline src/scancode/cli.py]: Noted.

**Comment by kartiksibal** [comment]
@pombredanne I'll re-structure it as you have guided. I have a small query. 
 
If suppose, our objective was to go ahead with my approach. 
Have I coded it, in an efficient way, then? In what way could I have changed my approach angle. You don't have to elaborate, just a subtle hint will be awesome! _Apart from the mistake on the file closing_. 

PS: I do realise the code is not enough to judge or comment. But, I'd love any input.
  
Thanks

**Comment by pombredanne** [comment]
@kartiksibal what is missing in all cases is tests first and then some docs. 

- Tests are needed to set the specs of your code in stone: I would have no idea how to use your approach as it is and what it means and what you expect it to be. Tests is the doc you write for the other developers.
- Doc is needed to explain a user-visible feature: a user would likely be puzzled as to why a `*_log.txt` would have popped up in the scanned code directory and what it meant. Rescanning the same directory a second time would yield different results as your log file would be now scanned that second time.

So the net net: when you write code, there are two goals:
1. your code will be read eventually 100 of times by other devs.  Therefore, you write first for your code it to be red and understood first. Hence the need for tests and at times some comments for non-obvious parts.
2. your code will be used eventually 100 of times by users.Therefore your need a user experience and documentation that makes it easy to understand, as self-evident as possible and non-surprising.

**Comment by pombredanne** [comment]
As a side note your use of a `patch-1` branch suggests that you may have used the online Github editor for this. This is in practice OK for small trivial typo edits. I do not think this is something that works for anything non-trivial involving multiple files, tests and more such as eventual amend and squash that may be needed. There are (many) better tools out there to edit code locally and the git CLI and GUI tools are your friends there.

**Comment by kartiksibal** [comment]
@pombredanne That's just the advise I needed! 
I had actually thought of adding tests later, once you were satisfied with my approach. I'll make sure I add them hand in hand.

I am completely aware about the online editor, used it only since the change was minuscule. :-)

Thank You so much! :grin:

**Comment by pombredanne** [comment]
So in terms of the approach, we cannot use a log file. We need this as some history list in the scan header instead. Can you rework that along these lines? Thanks!

**Comment by kartiksibal** [comment]
@pombredanne Sure, I'll get onto it once the application period is over?

File: src/scancode/cli.py (hunk 2 above)
Comment (kartiksibal) [inline src/scancode/cli.py]: @pombredanne I digged a little commoncode is somewhat eventually doing exactly this. What would you suggest?

**Comment by kartiksibal** [comment]
FML! 
Would you mind if I close this PR and start a new one.

File: src/scancode/cli.py (hunk 2 above)
Comment (pombredanne) [inline src/scancode/cli.py]: tell me what you found there. I am sure I wrote such a function in the past in there

Comment (kartiksibal) [inline src/scancode/cli.py]: In `commoncode/timeutils.py`, there's a function `tstamp2time` which takes a UTC time stamp as input and converts to a datetime object & a function `time2tstamp` which returns a UTC time stamp. So, if we feed the latter to the former. It'll return a timestamp. 
In such a case, is it better to call `commoncode`?

**Comment by pombredanne** [comment]
>  Would you mind if I close this PR and start a new one.

Sure. Make sure you reference this PR in the body of the new PR.... for reference

**Comment by pombredanne** [comment]
@kartiksibal I am closing this as requested and will wait for the new one instead.

**Comment by kartiksibal** [comment]
@pombredanne Shoot! I got too engrossed in the project. 😅 
Will come back to the issues, now. 😄

**Comment by pombredanne** [comment]
feel free to reopen this one if you want

**Pull request closed without merging**